FLT/DedekindDomain/AdicValuation.lean

```diff
@@ -0,0 +1,288 @@
+  rw [denseRange_iff_closure_range]
+  ext x
+  simp only [Set.mem_univ, iff_true]
+  rw [closure_subtype]
+  suffices h : Subtype.val '' Set.range ((algebraMap A ↥(HeightOneSpectrum.adicCompletionIntegers K v))) =
+      (algebraMap A (v.adicCompletion K)).range by
+    rw [h, closureAlgebraMapIntegers_eq_integers K v]
+    exact Subtype.coe_prop x
+  simp only [RingHom.coe_range, ← Set.range_comp']
+  rfl
+
+/-- An element of `𝒪_v` can be approximated by an element of `A`. -/
+theorem exists_adicValued_sub_lt_of_adicCompletionInteger ( x : v.adicCompletionIntegers K )
+    ( γ : (WithZero (Multiplicative ℤ))ˣ ) :
+    ∃a, Valued.v ((algebraMap A K a) - (x : v.adicCompletion K)) < γ.val := by
+  have h := closureAlgebraMapIntegers_eq_integers K v
+  rw [Set.ext_iff] at h
+  specialize h x
+  simp_rw [RingHom.coe_range, Subtype.coe_prop, iff_true, mem_closure_iff_nhds] at h
+  specialize h { y | Valued.v (y  - (x : v.adicCompletion K)) < γ.val }
+  have hn : {y | Valued.v (y - (x : v.adicCompletion K)) < γ.val} ∈ nhds x.val := by
+    rw [Valued.mem_nhds]
+    use γ
+  obtain ⟨z, ⟨hz, a, ha⟩⟩ := h hn
+  use a
+  rw [HeightOneSpectrum.algebraMap_adicCompletion, Function.comp_apply] at ha
+  rwa [ha]
+
+/-- An element of `∏_{v ∈ s} 𝒪_v`, with `s` finite, can be approximated by an element of `A`.
+  -/
+theorem exists_forall_adicValued_sub_lt {ι : Type*} (s : Finset ι)
+    (e : ι → (WithZero (Multiplicative ℤ))ˣ ) (valuation : ι → HeightOneSpectrum A)
+    (injective : Function.Injective valuation)
+    (x : (i : ι) → (valuation i).adicCompletionIntegers K) :
+    ∃ a, ∀ i ∈ s, Valued.v ((algebraMap A K a) - (x i).val) < (e i).val := by
+  -- Approximate the completion integers with actual integers using the previous lemma
+  let ⟨f, hf⟩ := Classical.axiomOfChoice
+    fun (i : s) => (exists_adicValued_sub_lt_of_adicCompletionInteger K (valuation i) (x i) (e i))
+
+  -- Convert the hypotheses from being about valuations to being about ideals, so
+  -- that we can apply (a suitable corollary of) the Chinese remainder theorem.
+  let e' : ι → ℕ := fun i => (WithZero.unzero (e i).ne_zero)⁻¹.toNat + 1
+  have he' : ∀ (i : ι), (Multiplicative.ofAdd (-(e' i : ℤ))) < (e i).val := by
+    intro i
+    unfold e'
+    simp only [Nat.cast_add, Int.ofNat_toNat, Nat.cast_one, neg_add_rev, Int.reduceNeg, ofAdd_add,
+      ofAdd_neg, WithZero.coe_mul, WithZero.coe_inv]
+    have mul_lt_of_lt_one_of_le' {a b c : WithZero (Multiplicative ℤ)} (ha : a < 1) (bc : b ≤ c)
+      (a0 : 0 ≤ a) (c0 : 0 < c) : a * b < c :=
+      (mul_le_mul_of_nonneg_left bc a0).trans_lt <| mul_lt_of_lt_one_left c0 ha
+    apply mul_lt_of_lt_one_of_le'
+    . norm_cast
+    . rw [← WithZero.coe_inv]
+      apply le_of_le_of_eq _ (WithZero.coe_unzero (e i).ne_zero)
+      rw [WithZero.coe_le_coe, ← ofAdd_neg, neg_sup]
+      apply inf_le_of_left_le
+      apply le_of_eq (Int.neg_neg (WithZero.unzero (e i).ne_zero))
+    . norm_cast
+    apply Units.zero_lt
```

Comment:
You've proved this before, right? This is again just the statement that given an element of Z_m (the nonzero elements of Z_m0), there's an element of -N_m which is less than it. This proof looks even longer than the one you gave the first time, when you proved that given two elements of Z_m there was an element of -N_m which is less than both of them. I think you should just pull out this lemma as a sublemma and just use it both here and in the two-variable case.